clients/src/test/java/org/apache/kafka/common/config/ConfigDefTest.java

```diff
@@ -164,6 +170,91 @@ public void testNullDefaultWithValidator() {
         assertEquals("ONE", vals.get(key));
     }
 
+    @Test
+    public void testValidate() {
+        Map<String, ConfigValue> expected = new HashMap<>();
+        String errorMessageB = "Missing required configuration \"b\" which has no default value.";
+        String errorMessageC = "Missing required configuration \"c\" which has no default value.";
+        String errorMessageD = "Invalid value for configuration d";
+
+        ConfigValue configA = new ConfigValue("a", 1, Arrays.<Object>asList(1, 2, 3), Collections.<String>emptyList());
+        ConfigValue configB = new ConfigValue("b", null, Arrays.<Object>asList(4, 5), Arrays.asList(errorMessageB, errorMessageB));
+        ConfigValue configC = new ConfigValue("c", null, Arrays.<Object>asList(4, 5), Arrays.asList(errorMessageC));
+        ConfigValue configD = new ConfigValue("d", 10, Arrays.<Object>asList(1, 2, 3), Arrays.asList(errorMessageD));
+
+        expected.put("a", configA);
+        expected.put("b", configB);
+        expected.put("c", configC);
+        expected.put("d", configD);
+
+        ConfigDef def = new ConfigDef()
+            .define("a", Type.INT, Importance.HIGH, "docs", "group", 1, Width.SHORT, "a", Arrays.asList("b", "c"), new IntegerRecommender(false))
+            .define("b", Type.INT, Importance.HIGH, "docs", "group", 2, Width.SHORT, "b", new IntegerRecommender(true))
+            .define("c", Type.INT, Importance.HIGH, "docs", "group", 3, Width.SHORT, "c", new IntegerRecommender(true))
+            .define("d", Type.INT, Importance.HIGH, "docs", "group", 4, Width.SHORT, "d", Arrays.asList("b"), new IntegerRecommender(false));
+
+        Map<String, String> props = new HashMap<>();
+        props.put("a", "1");
+        props.put("d", "10");
+
+        List<ConfigValue> configs = def.validate(props);
+        for (ConfigValue config : configs) {
+            String name = config.name();
+            ConfigValue expectedConfig = expected.get(name);
+            verifyConfigValues(expectedConfig, config);
+        }
+    }
+
+    private void verifyConfigValues(ConfigValue a, ConfigValue b) {
+        assertEquals(a.name(), b.name());
+
+        if (a.value() != null && b.value() != null) {
+            int aValue = (int) a.value();
+            int bValue = (int) b.value();
+            assertEquals(aValue, bValue);
+        }
+
+        List<Object> aRecommendedValues = a.recommendedValues();
+        List<Object> bRecommendedValues = b.recommendedValues();
+        assertEquals(aRecommendedValues.size(), bRecommendedValues.size());
```

Comment:
Why only check their size, why not also check contents of the lists? Can this entire method be replaced just by checking equality of the two `ConfigValue` objects?